Fixes #1084: Tablet: Highfi Onboarding Flow (landscape)

MohamedMedhat1998 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
- Fixes #1084
- Added layouts for `Onboarding` screens for `Tablet` devices in `Landscape` mode (**High-fi**  )
- Mock link: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/
## Screenshots
![screenshot-2020-05-30_12 37 04 434](https://user-images.githubusercontent.com/8533363/83326244-a2d41180-a272-11ea-8465-5833162c29d4.png)
![screenshot-2020-05-30_12 37 12 047](https://user-images.githubusercontent.com/8533363/83326246-a798c580-a272-11ea-9cea-eed445568dfa.png)
![screenshot-2020-05-30_12 37 17 295](https://user-images.githubusercontent.com/8533363/83326247-abc4e300-a272-11ea-946e-aa85bc344a97.png)
![screenshot-2020-05-30_12 37 23 743](https://user-images.githubusercontent.com/8533363/83326248-aebfd380-a272-11ea-85bd-0a2228cdfdc3.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

MohamedMedhat1998 committed:
-Onboarding: draft implementation for tablet-landscape high-fi

MohamedMedhat1998 committed:
-Removed the unused drawable file
-Used `app:layout_constraintDimensionRatio="H, 10:9"` and `android:scaleType="fitXY"` instead of `guideline` and `android:scaleType="centerCrop"`
-Centered the `textViews` between the `imageView` and the end of the `parentLayout` and gave them a fixed width of `360dp` instead of placing the `textViews` relative to the end of the `parentLayout` only

MohamedMedhat1998 committed:
-placed the textViews and the buttons relative to the end of the screen (with a margin of 96dp) and a guideline of 57% from the beginning of the screen instead of using fixed width of 360dp

rt4914 commented on app/src/main/res/drawable-sw600dp-land/onboarding_dot_active.xml:
```diff
@@ -0,0 +1,6 @@
+  android:shape="oval">
+  <solid android:color="@color/oppiaPrimaryText" />
+  <corners android:radius="5dp" />
+</shape>
```
Remove this file and use the existing `onboarding_dot_active`.
We normally round up the dimensions to nearest dimension `4dp` and as it is `5px` in mocks we can use `4dp` here.

All dimensions should be in multiple of `4dp`. (There are some exceptions like border strokes)

MohamedMedhat1998 replied:
Done

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_fragment.xml:
```diff
@@ -95,8 +96,9 @@
+          app:layout_constraintBottom_toBottomOf="@+id/slide_dots_container"
           app:layout_constraintStart_toStartOf="parent"
           app:layout_constraintTop_toTopOf="@+id/slide_dots_container" />
       </androidx.constraintlayout.widget.ConstraintLayout>
```
The parent of this viewgroup which is `FrameLayout` is actually doing nothing for these mocks, so you can remove that `FrameLayout` and use `ConstraintLayout` directly here.

MohamedMedhat1998 replied:
Done

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_fragment.xml:
```diff
@@ -95,8 +96,9 @@
           android:visibility="@{viewModel.slideNumber != (viewModel.totalNumberOfSlides - 1) ? View.VISIBLE: View.GONE, default=visible}"
+          app:layout_constraintBottom_toBottomOf="@+id/slide_dots_container"
           app:layout_constraintStart_toStartOf="parent"
           app:layout_constraintTop_toTopOf="@+id/slide_dots_container" />
```
Color code for `onboarding_fragment_divider_view ` is `#707070` in the mocks and you are using `#333333` which is incorrect.

MohamedMedhat1998 replied:
Done

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_fragment.xml:
```diff
@@ -70,23 +70,24 @@
           android:minHeight="48dp"
           android:onClick="@{(v) -> presenter.clickOnNext()}"
-          android:visibility="@{viewModel.slideNumber != 3 ? View.VISIBLE: View.GONE, default=visible}"
           android:paddingStart="12dp"
```
Remove this padding start and padding end.

Instead you have to use `marginStart = 12dp` and `marginEnd = 12dp`

Also, just an explanation of how you can easily cross-verify out these things

The bottom bar layout height is `72dp` in mocks.
The height of image button arrow is `48dp`
The margin top for arrow is `12dp`as per mocks
The margin bottom for arrow is `12dp` as per mocks

72dp = 48dp + 12dp + 12dp

MohamedMedhat1998 replied:
Done

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_slide_final.xml:
```diff
@@ -18,9 +18,9 @@
+      android:scaleType="centerCrop"
       android:src="@drawable/ic_landscape_onboarding_3"
       app:layout_constraintBottom_toBottomOf="parent"
       app:layout_constraintDimensionRatio="H, 10:9"
```
As per mocks the dimension of image is `640x624`  (39:40 ratio)
And as per image file in drawable the dimensions are `320x360` (9:8 ratio)

So either the mocks should be updated to us a ratio of `9:8` (Maybe use 555 px width of image `555 = 624*8/9`)


Or we should have image which has dimension of `640x624` which can be used for tablet layouts.

@mschanteltc PTAL

mschanteltc replied:
The images in the mocks start from the top of the screen and end above the navbar. This means that some of the image is below the sysbar and the page navigation bar (Skip, Page Indicators, Right Arrow). Would it be possible that we can overlap these or would a cropped image work better?

rt4914 replied:
@mschanteltc Cropped version would actually work better here.

rt4914 replied:
@mschanteltc Can you please provide these images? Cropped images.

mschanteltc replied:
Created a new [image folder](https://drive.google.com/drive/folders/1ph16B-BPOvixkuL31hI6cIB8F0HSEx6Y?usp=sharing) with updated images to be cropped.

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_slide_final.xml:
```diff
@@ -85,5 +79,12 @@
       android:orientation="horizontal"
       app:layout_constraintGuide_percent="0.5" />
+
+    <androidx.constraintlayout.widget.Guideline
```
Please don't use guidlines, as for the way we have prepared mocks, it does not guarantee the at a fix percentage the text will start or end. Avoid using guidlines in such cases and instead rely on margin/padding as they provide more robust solution and can work across devices.

MohamedMedhat1998 replied:
But if I'm going to remove the `guidlines`, I'll have to use some fixed widths.

rt4914 replied:
Why do we need fixed width if we remove guidelines? I mean, we have done in mobile screens without guidelines and it works fine in HDPI + Screens.

Can you please explain how will it affect the UI if we do not use guidelines and fixed width and instead simply use margins/paddings.

MohamedMedhat1998 replied:
I mean that if we didn't use guidelines nor fixed widths, we'll have some options here:  
1. Make a constraint between the `TextView` and the `Thumbnail` and between the `TextView` and the `Parent` then set the width of the `TextView` to `match_constraint` (0dp) and use the suitable margins
2. Make only 1 constraint between the `TextView` and the `Parent` and use the suitable margins and make the width of the `TextView` to `wrap_content`
And in both options, the text in each `TextView` will be only in one line, but in the mocks we have multiple lines.  
So, to have multiple lines like the mocks, I think we'll have to use either guidlines or fixed widths

rt4914 replied:
Okay, then in that case we will need the guidelines. Lets get the images from Chantel and then update this accordingly.

rt4914 reviewed (changes requested):
PTAL

MohamedMedhat1998 committed:
-Deleted "onboarding_dot_active.xml" with "sw600dp-land" qualifier
-Removed the unnecessary "FrameLayout" in "onboarding_fragment.xml"
-Removed the "guidelines" in "onboarding_slide.xml" and "onboarding_slide_final.xml" and replaced them with margins and fixed widths
-Updated the "oppiaOnboardingDivider" in "colors.xml" and used it in the "onboarding_fragment.xml"

MohamedMedhat1998 committed:
-Updated the UI

MohamedMedhat1998 commented:
@rt4914 PTAL

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_slide.xml:
```diff
@@ -16,49 +16,43 @@
       android:scaleType="fitXY"
       android:src="@{viewModel.slideImage}"
       app:layout_constraintBottom_toBottomOf="parent"
-      app:layout_constraintDimensionRatio="H, 10:9"
```
Keep this. This ratio makes sure that image is always visible in best way.

MohamedMedhat1998 replied:
After following this approach -and setting the width to 0dp-, the thumbnails appeared horizontally compressed.  
## After using 
```
app:layout_constraintDimensionRatio="H, 10:9"
```
![screenshot-2020-05-31_11 12 41 093](https://user-images.githubusercontent.com/8533363/83348801-0c662580-a330-11ea-8eda-1654f5cccb60.png)
![screenshot-2020-05-31_11 12 55 522](https://user-images.githubusercontent.com/8533363/83348825-3ddef100-a330-11ea-896a-20b560c9035c.png)
## Before using
```
app:layout_constraintDimensionRatio="H, 10:9"
```
![screenshot-2020-05-30_12 37 12 047](https://user-images.githubusercontent.com/8533363/83348805-11c37000-a330-11ea-811a-e8ae372eb711.png)
![screenshot-2020-05-30_12 37 23 743](https://user-images.githubusercontent.com/8533363/83348841-4e8f6700-a330-11ea-9e14-4e934444194e.png)

And it is obvious that sticking with the approach that doesn't use `app:layout_constraintDimensionRatio="H, 10:9"` will give us a better result as it looks exactly the same as the [mocks](https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/00a93a92-aa03-42d4-9a98-d186eaeec2ca/L-Onboarding-4). But using `app:layout_constraintDimensionRatio="H, 10:9"` will make the images a bit compressed.  
The first approach will always look better as it uses `android:layout_height="0dp"` so that the image will always fill the screen vertically, and `android:layout_width="wrap_content"` which will make the image take whatever width it needs, this will make the image appear with the correct ratio because the image file itself is in the correct ratio. Finally, `android:scaleType="fitXY"` ensures that the image is in the right place.
@rt4914 PTAL

rt4914 replied:
Try `40:39` or `39:40` that should work because the image that you are using here follows that ratio.

Ratios are important ass they make sure that behaviour is consistent across devices. In this Screen this looks correct because mocks have been designed for this only.

MohamedMedhat1998 replied:
40:39 is perfect 👌  
![screenshot-2020-05-31_14 34 29 311](https://user-images.githubusercontent.com/8533363/83352492-0b42f180-a34c-11ea-98db-99d2794fed8e.png)
![screenshot-2020-05-31_14 35 16 889](https://user-images.githubusercontent.com/8533363/83352498-1433c300-a34c-11ea-9ec7-af541fd50704.png)
![screenshot-2020-05-31_14 35 40 305](https://user-images.githubusercontent.com/8533363/83352501-185fe080-a34c-11ea-93e8-c7fbe909c167.png)
![screenshot-2020-05-31_14 35 00 609](https://user-images.githubusercontent.com/8533363/83352496-10a03c00-a34c-11ea-9853-b5e593f0cd0e.png)

rt4914 replied:
Thanks.

rt4914 commented on app/src/main/res/layout-sw600dp-land/onboarding_slide_final.xml:
```diff
@@ -16,67 +16,59 @@
-      android:src="@drawable/ic_landscape_onboarding_3"
+      android:src="@drawable/ic_landscape_onboarding_3_tablet"
       app:layout_constraintBottom_toBottomOf="parent"
-      app:layout_constraintDimensionRatio="H, 10:9"
```
Ditto as above.

rt4914 replied:
@MohamedMedhat1998 Please make sure that you reply to each comment. Thanks.

MohamedMedhat1998 replied:
Ok  
Followed the same approach
> 
> 
> 40:39 is perfect 👌
> ![screenshot-2020-05-31_14 34 29 311](https://user-images.githubusercontent.com/8533363/83352492-0b42f180-a34c-11ea-98db-99d2794fed8e.png)
> ![screenshot-2020-05-31_14 35 16 889](https://user-images.githubusercontent.com/8533363/83352498-1433c300-a34c-11ea-9ec7-af541fd50704.png)
> ![screenshot-2020-05-31_14 35 40 305](https://user-images.githubusercontent.com/8533363/83352501-185fe080-a34c-11ea-93e8-c7fbe909c167.png)
> ![screenshot-2020-05-31_14 35 00 609](https://user-images.githubusercontent.com/8533363/83352496-10a03c00-a34c-11ea-9853-b5e593f0cd0e.png)

rt4914 replied:
Great. Thanks.

rt4914 commented on app/src/main/res/values/booleans.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <bool name="isTablet">false</bool>
+</resources>
```
EOF is missing, add one blank line at the end.

MohamedMedhat1998 replied:
Done.

rt4914 commented on app/src/main/res/values-sw600dp/booleans.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <bool name="isTablet">true</bool>
+</resources>
```
EOF is missing, add one blank line at the end.

MohamedMedhat1998 replied:
Done.

rt4914 reviewed (changes requested):
Looks good. Final nit changes suggested.

MohamedMedhat1998 committed:
-Added EOF in "booleans.xml"

MohamedMedhat1998 committed:
-Updated the ratio of the thumbnail images to be "40:39"

rt4914 reviewed (approved):
LGTM, just waiting for your reply on one comment that I posted earlier.

MohamedMedhat1998 committed:
-Changed the height of the "Get Started" button to be "wrap_content" (minHeight is set to 48dp)

MohamedMedhat1998 commented:
> 
> 
> LGTM, just waiting for your reply on one comment that I posted earlier.

I have added a new commit that follows instructions mentioned in [this comment](https://github.com/oppia/oppia-android/pull/1153#discussion_r433068838). Which is changing the height from fixed height to `wrap_content` with `minHeight="48dp"`

rt4914 reviewed (approved):
LGTM, thanks.

rt4914 commented:
@MohamedMedhat1998 Please merge your branch with latest develop yo avoid falky_tests run.

MohamedMedhat1998 committed:
Merge remote-tracking branch 'origin/develop' into onboarding-tablet-land-highfi

MohamedMedhat1998 commented:
> 
> 
> @MohamedMedhat1998 Please merge your branch with latest develop yo avoid falky_tests run.

@rt4914 Done

MohamedMedhat1998 committed:
Merge remote-tracking branch 'origin/develop' into onboarding-tablet-land-highfi

MohamedMedhat1998 committed:
Merge remote-tracking branch 'origin/develop' into onboarding-tablet-land-highfi
